### PR TITLE
Fix datetimepicker clear date selection

### LINKED ISSUE #466: Cannot clear selected Date in DateTimePicker
### Bug report

- __Package version(s)__: dateTime 1.4.0
- __Browser and OS versions__: Chrome 55.0.2883.87

#### Steps to reproduce

1. Add DateTimePicker with the ability to clear selection
```jsx
<DateTimePicker
  datePickerProps={{
    canClearSelection: true,
    showActionsBar: true,
  }}
/>
```
2. Click Clear

#### Actual behavior

`Uncaught TypeError: Cannot read property 'getFullYear' of null`
`getDateTime` in common/DateUtils immediately calls `date.getFullYear()` on its `date` argument, however the value passed to it by `DateTimePicker.handleDateChange` is `null` when the user clicks the clear button.

#### Expected behavior

The plugin clears the selected date and passes the change (value `null`) to the `onChange` DateTimePicker prop.

### PR DESCRIPTION
**Pull request by leebyp**
#### Checklist

- [x] Include tests

#### Changes proposed in this pull request:

Split up the internal state in `DateTimePicker` to let the date and time be manipulated separately.

#### Reviewers should focus on:

I considered keeping `value` in the state in addition to `dateValue` and `timeValue` to quickly access the combination of the two. Would this be more confusing than it is useful?

### CODE REVIEW

File: packages/datetime/src/common/dateUtils.ts [hunk 1]
```diff
@@ -98,6 +98,12 @@ export function getDateBetween(dateRange: DateRange) {
 }
 
 export function getDateTime(date: Date, time: Date) {
-    return new Date(date.getFullYear(), date.getMonth(), date.getDate(),
+    if (date === null) {
+        return null;
```
Comment (adidahiya) [inline packages/datetime/src/common/dateUtils.ts]: can we return `undefined` instead? prefer avoiding explicit nulls

Comment (leebyp) [inline packages/datetime/src/common/dateUtils.ts]: This is used for the `DatePicker` within `DateTimePicker` and `null` / `undefined` will differentiate between controlled / uncontrolled.

File: packages/datetime/src/common/dateUtils.ts [hunk 2]
```diff
@@ -98,6 +98,12 @@ export function getDateBetween(dateRange: DateRange) {
...
+    } else if (time === null) {
+        return new Date(date.getFullYear(), date.getMonth(), date.getDate(), 0, 0, 0, 0);
+    } else {
+        return new Date(date.getFullYear(), date.getMonth(), date.getDate(),
            time.getHours(), time.getMinutes(), time.getSeconds(), time.getMilliseconds());
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: indentation is off here. should be +1 or +5

File: packages/datetime/src/dateTimePicker.tsx [hunk 3]
```diff
@@ -60,46 +62,52 @@ export class DateTimePicker extends AbstractComponent<IDateTimePickerProps, IDat
...
+            this.setState({
+                dateValue: nextProps.value,
+                timeValue: nextProps.value,
+            });
         }
     }
 
     public handleDateChange = (date: Date, isUserChange: boolean) => {
-        const value = DateUtils.getDateTime(date, this.state.value);
+        const value = DateUtils.getDateTime(date, this.state.timeValue);
```
Comment (giladgray) [inline packages/datetime/src/dateTimePicker.tsx]: merge the objects only for invoking `onChange`. store `date` arg directly in state.

cuz `dateValue` doesn't care at all about its time components and vice-versa for `timeValue` so why bother merging them?

Comment (leebyp) [inline packages/datetime/src/dateTimePicker.tsx]: that was causing another bug too, fixed and added test

**Pull request merged**